### PR TITLE
[Android] Added props to Collapsing Top Bar

### PR DESCRIPTION
**Pull request by hanswage**
The original Collapsable Top Bar does not allow to add props to the collapsing top bar component, but does allow the usage of a React Native component.

In the static navigatorStyle value, add:
`
collapsingToolBarComponent: 'example.Transitions.UpdatableCollapsingHeader.component',
collapsingToolBarComponentInitialProps: {costs: 10},
`

In our subview, we can now retrieve the 'costs' value by calling this.props.costs.

To update this value, use:
`
this.props.navigator.setStyle({
  collapsingToolBarComponentInitialProps: {costs: this.state.costs},
});
`

I have added an example to the example project. It is my first commit to another project, so feel free to comment or change.

### CODE REVIEW

File: android/app/src/main/java/com/reactnativenavigation/views/collapsingToolbar/CollapsingTopBar.java [hunk 1]
```diff
@@ -47,6 +48,21 @@ public CollapsingTopBar(Context context, final StyleParams params) {
...
 
+    @Override
+    public void setStyle(StyleParams styleParams) {
+        super.setStyle(styleParams);
+
+        if (header != null &&
+            header.getScreenId() != null &&
+            !header.getScreenId().equals(styleParams.collapsingTopBarParams.reactViewId)) {
+            // The user provided a new collapsing top bar react view
+            createReactHeader(styleParams);
```
[the remark this answers is not in the document]
Comment (guyca) [reply android/app/src/main/java/com/reactnativenavigation/views/collapsingToolbar/CollapsingTopBar.java]: I've never used the `setAppProperties` api, if it's called after a component is already created, does it "restart" it (will the component go through all of its react life cycle methods?)

File: android/app/src/main/java/com/reactnativenavigation/views/collapsingToolbar/CollapsingTopBar.java [hunk 2]
```diff
@@ -75,8 +91,13 @@ private void createBackgroundImage() {
         }
     }
 
-    private void createReactHeader(CollapsingTopBarParams params) {
-        if (params.hasReactView()) {
+    private void createReactHeader(StyleParams params) {
```
Comment (guyca) [inline android/app/src/main/java/com/reactnativenavigation/views/collapsingToolbar/CollapsingTopBar.java]: Why was the parameter changed from `CollapsingTopBarParams` to `StyleParams` ?

File: android/app/src/main/java/com/reactnativenavigation/views/collapsingToolbar/CollapsingTopBarReactHeader.java [hunk 3]
```diff
@@ -45,6 +45,11 @@ public CollapsingTopBarReactHeader(Context context, CollapsingTopBarParams param
         setOnDisplayListener(this);
     }
 
+    public void setStyle(StyleParams styleParams) {
+        // If the setAppProperties method cannot be found, please update React Native
+        setAppProperties(styleParams.collapsingTopBarComponentInitialProps);
```
Comment (guyca) [inline android/app/src/main/java/com/reactnativenavigation/views/collapsingToolbar/CollapsingTopBarReactHeader.java]: Please add collapsingTopBarComponentInitialProps to CollapsingTopBarParams

File: android/app/src/main/java/com/reactnativenavigation/views/collapsingToolbar/CollapsingTopBarReactHeader.java [hunk 4]
```diff
@@ -35,8 +35,8 @@ public void setOnHiddenListener(CollapsingTopBarReactHeaderAnimator.OnHiddenList
         this.onHiddenListener = onHiddenListener;
     }
 
-    public CollapsingTopBarReactHeader(Context context, CollapsingTopBarParams params, NavigationParams navigationParams, ScrollListener scrollListener, Screen.OnDisplayListener onDisplayListener) {
-        super(context, params.reactViewId, navigationParams);
+    public CollapsingTopBarReactHeader(Context context, StyleParams params, NavigationParams navigationParams, ScrollListener scrollListener, Screen.OnDisplayListener onDisplayListener) {
```
Comment (guyca) [inline android/app/src/main/java/com/reactnativenavigation/views/collapsingToolbar/CollapsingTopBarReactHeader.java]: Lets continue using CollapsingTopBarParams if possible

File: android/gradle/wrapper/gradle-wrapper.properties [hunk 5]
```diff
@@ -1,6 +1,6 @@
-#Mon Sep 12 16:05:44 IDT 2016
+#Wed Aug 30 14:35:19 CEST 2017
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-2.14.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-2.14.1-all.zip
```
Comment (guyca) [inline android/gradle/wrapper/gradle-wrapper.properties]: This change isn't related to the PR

File: package.json [hunk 6]
```diff
@@ -52,8 +52,8 @@
     "jest-cli": "18.x.x",
     "jest-react-native": "18.x.x",
     "prop-types": "^15.5.10",
-    "react": "16.0.0-alpha.6",
-    "react-native": "0.43.0",
+    "react": "16.0.0-alpha.12",
+    "react-native": "0.45.1",
```
Comment (guyca) [inline package.json]: Can we keep using RN 0.43 for now?

File: yarn.lock [hunk 7]
```diff
@@ -208,12 +208,22 @@ async@^1.4.0:
   version "1.5.2"
   resolved "https://registry.yarnpkg.com/async/-/async-1.5.2.tgz#ec6a61ae56480c0c3cb241c95618e20892f9672a"
 
-async@^2.0.1, async@^2.1.4:
+async@^2.1.4:
```
Comment (guyca) [inline yarn.lock]: Please don't commit changes to yarn.lock. It shouldn't be committed

**Comment by guyca** [comment]
Hey @hanswage 
Thanks a lot for the PR. I see a lot of effort was put into it!

I'm don't think calling `setStyle` to update props for a component is the right thing to do. Passing initial props was a compromise we had to do. The header component should be a connected component and updating its props should be done via flux or perhaps [deeplinks](https://wix.github.io/react-native-navigation/#/deep-links)

**Comment by exentrich** [comment]
@hanswage can you update your pr please?

**Pull request closed without merging**